### PR TITLE
*: refine coordinator startup

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 

 coordinator startup:

- only leader startup 
- recover the registered workers and their binding relationship,

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: dm/master/election.go [hunk 1]
```diff
@@ -30,13 +30,19 @@ func (s *Server) electionNotify(ctx context.Context) {
 			// output the leader info.
 			if leader {
 				log.L().Info("current member become the leader", zap.String("current member", s.cfg.Name))
+				err := s.coordinator.Start(ctx, s.etcdClient)
```
Comment (csuzhangxc) [inline dm/master/election.go]: If before the old leader called `s.coordinator.Stop()`, the new leader called `s.coordinator.Start`, is it OK?

File: dm/worker/server.go [hunk 2]
```diff
@@ -215,7 +228,17 @@ func (s *Server) StartSubTask(ctx context.Context, req *pb.StartSubTaskRequest)
...
+		ctx, cancel := context.WithTimeout(s.etcdClient.Ctx(), 3*time.Second)
+		defer cancel()
+		_, err = s.etcdClient.Put(ctx, common.UpstreamSubTaskKeyAdapter.Encode(s.cfg.WorkerAddr), cfg.String())
+		if err != nil {
+			resp.Result = false
+			resp.Msg = err.Error()
+		}
 	}
+	// FIXME: handle error
+	_ = w.OperateSubTask(cfg.Name, pb.TaskOp_Stop)
```
Comment (csuzhangxc) [inline dm/worker/server.go]: `TaskOp_Stop` 🤔

File: dm/master/election.go (hunk 1 above)
Comment (nolouch) [reply dm/master/election.go]: I think we can add a validator before RPC to handle that check if the dm-master is a leader. that may be related to the gRPC proxy.

Comment (csuzhangxc) [reply dm/master/election.go]: OK

File: dm/master/coordinator/coordinator.go [hunk 3]
```diff
@@ -55,9 +63,87 @@ func NewCoordinator() *Coordinator {
...
+
+	for _, kv := range resp.Kvs {
+		addr := common.UpstreamBoundWorkerKeyAdapter.Decode(string(kv.Key))[0]
+		sourceID := string(kv.Value)
+		w, ok := c.workers[addr]
+		if !ok {
+			log.L().Error("worker not exist but binding relationship exist", zap.String("addr", addr), zap.String("source", sourceID))
+			continue
+		}
+		resp, err = etcdClient.Get(ctx, common.UpstreamConfigKeyAdapter.Encode(sourceID))
```
Comment (csuzhangxc) [inline dm/master/coordinator/coordinator.go]: Does this `resp` overwrite the value of it in L96?

Comment (nolouch) [reply dm/master/coordinator/coordinator.go]: actually `for ... range` will clone the values, not affect the origin value.

**Pull request merged**